trie.h

```diff
@@ -11,6 +11,30 @@
 //////////////////////////////////////////////////////////////////
 // Public API
 
+typedef struct
+{
+    uint16_t    completion_index;   // index to start of completion string in trie_t.completions
+    uint8_t     completion_len;     // length of completion string
+    uint8_t     num_backspaces;     // number of backspaces to send before the completion string
+    uint8_t     func_code;          // special function code
+} st_trie_payload_t;
+
+typedef struct
+{
+    int     pos;                // buffer index of cursor position
+    int     sub_pos;            // Sub-position within the current buffer position
+    int     segment_len;        // Number of elements traversed
+    uint8_t as_output_buffer;   // True if buffer traversing the simulated output
+} st_cursor_pos_t;
+
+typedef struct
+{
+    st_key_buffer_t         *buffer;            // buffer
+    st_cursor_pos_t         cursor_pos;         // Contains all position info for the cursor
+    st_trie_payload_t       cached_action;
+    uint8_t                 cache_valid;
```

Comment:
should be bool if this is used as true/false

Reply:
using uint8_t to save stack space.